Lantern Network Setup/Lantern Network Setup/main.m

```diff
@@ -1,102 +1,134 @@
-        runNetworkSetup(@"-setautoproxystate", s, onOff);
+void toggleAutoProxyConfigFile(NSString *onOff, NSString *autoProxyConfigFileUrl)
+{
+    
+    NSLog(@"Toggle %@ auto proxy configuration file to %@", onOff, autoProxyConfigFileUrl);
+    BOOL success = FALSE;
+    
+    SCNetworkSetRef networkSetRef;
+    CFArrayRef networkServicesArrayRef;
+    SCNetworkServiceRef networkServiceRef;
+    SCNetworkProtocolRef proxyProtocolRef;
+    NSDictionary *oldPreferences;
+    NSMutableDictionary *newPreferences;
+    NSString *wantedHost;
+    
+    
+    // Get System Preferences Lock
+    SCPreferencesRef prefsRef = SCPreferencesCreate(NULL, CFSTR("org.getlantern.lantern"), NULL);
+    
+    if(prefsRef==NULL) {
+        NSLog(@"Fail to obtain Preferences Ref!!");
+        goto freePrefsRef;
     }
-
+    
+    success = SCPreferencesLock(prefsRef, TRUE);
+    if (!success) {
+        NSLog(@"Fail to obtain PreferencesLock");
+        goto freePrefsRef;
+    }
+    
+    // Get available network services
+    networkSetRef = SCNetworkSetCopyCurrent(prefsRef);
+    if(networkSetRef == NULL) {
+        NSLog(@"Fail to get available network services");
+        goto freeNetworkSetRef;
+    }
+    
+    //Look up interface entry
+    networkServicesArrayRef = SCNetworkSetCopyServices(networkSetRef);
+    networkServiceRef = NULL;
+    for (long i = 0; i < CFArrayGetCount(networkServicesArrayRef); i++) {
+        networkServiceRef = CFArrayGetValueAtIndex(networkServicesArrayRef, i);
+        NSLog(@"Setting proxy for device %@", SCNetworkServiceGetName(networkServiceRef));
+        
+        // Get proxy protocol
+        proxyProtocolRef = SCNetworkServiceCopyProtocol(networkServiceRef, kSCNetworkProtocolTypeProxies);
+        if(proxyProtocolRef == NULL) {
+            NSLog(@"Couldn't acquire copy of proxyProtocol");
+            goto freeProxyProtocolRef;
+        }
+        
+        oldPreferences = (__bridge NSDictionary*)SCNetworkProtocolGetConfiguration(proxyProtocolRef);
+        newPreferences = [NSMutableDictionary dictionaryWithDictionary: oldPreferences];
+        wantedHost = @"localhost";
+        
+        if([onOff  isEqual: @"on"]) {//Turn proxy configuration ON
+            [newPreferences setValue: wantedHost forKey:(NSString*)kSCPropNetProxiesHTTPProxy];
+            [newPreferences setValue:[NSNumber numberWithInt:1] forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigEnable];
+            [newPreferences setValue:autoProxyConfigFileUrl forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigURLString];
+            NSLog(@"Setting Proxy ON with: %@", newPreferences);
+        } else {//Turn proxy configuration OFF
+            [newPreferences setValue:[NSNumber numberWithInt:0] forKey:(NSString*)kSCPropNetProxiesProxyAutoConfigEnable];
+            NSLog(@"Setting Proxy OFF");
+        }
+        
+        success = SCNetworkProtocolSetConfiguration(proxyProtocolRef, (__bridge CFDictionaryRef)newPreferences);
```

Comment:
Will this overwrite any changes that other processes made between line 76 and here?